### ISSUE TITLE
sqlform.factory issue

### ISSUE THREAD

**Issue by mpranjic** [web2py/pydal#243]
It seems to me that sqlform factory is broken. Git bisect traced it down to commit @03dc3a8d4cddf3c1fd92cdab62d923d64f777776

```
Traceback (most recent call last):
  File "/home/marin/web2py/gluon/restricted.py", line 227, in restricted
    exec ccode in environment
  File "/home/marin/web2py/applications/move/controllers/user.py", line 142, in <module>
  File "/home/marin/web2py/gluon/globals.py", line 412, in <lambda>
    self._caller = lambda f: f()
  File "/home/marin/web2py/applications/move/controllers/user.py", line 98, in login
    form = SQLFORM.factory(Field('email', 'string'), Field('password', 'string')).process(keepvalues=True)
  File "/home/marin/web2py/gluon/html.py", line 2301, in process
    self.validate(**kwargs)
  File "/home/marin/web2py/gluon/html.py", line 2238, in validate
    if self.accepts(**kwargs):
  File "/home/marin/web2py/gluon/sqlhtml.py", line 1688, in accepts
    self.vars.id = self.table.insert(**fields)
  File "/home/marin/web2py/gluon/packages/dal/pydal/objects.py", line 706, in insert
    ret = self._db._adapter.insert(self, self._listify(fields))
  File "/home/marin/web2py/gluon/packages/dal/pydal/adapters/base.py", line 737, in insert
    raise e
ValueError: INSERT INTO no_table(password,email) VALUES ('password','test.marin@entropia.hr');
```

**Comment by stephenrauch** [comment]
Which database backend are you using?

**Comment by gi0baro** [comment]
I think is not related to pydal..

**Comment by mpranjic** [comment]
sqlite

**Comment by mpranjic** [comment]
> I think is not related to pydal..

Most likely it isn't but why it's triggered by commit in this repo?
It's not obvious to me.

**Comment by gi0baro** [comment]
The traceback points to pyDAL because it's receiving an insert instruction from SQLFORM factory with a fake table. In fact you see `no_table` in the traceback. Under my point of view this is a bug in `factory` which shouldn't try to insert values on non-existing tables.

**Comment by kjkuan** [comment]
just met the same problem I think. This used to work. Now I have to specify ``form.process(..., dbio=False)`` explicitly to work around this... :(

**Comment by gi0baro** [comment]
@ilvalle maybe something related to new classes you implemented? should we inspect this?
@mdipierro what do you think? maybe pyDAL has pointed out an hidden bug in `SQLFORM.factory`?

**Comment by ilvalle** [comment]
Can you post a failing example?

 Paolo

2015-07-03 1:23 GMT+02:00 Giovanni Barillari <notifications@github.com>:

**Comment by mpranjic** [comment]
Anyone knows why ```dbio``` is set to ```True``` [here](https://github.com/web2py/web2py/blob/master/gluon/html.py#L2299)?

Changing to the false seems to fix the problem (and looks obvious imo).

**Comment by mdipierro** [comment]
kwargs['dbio'] = kwargs.get('dbio', True)

only sets it true if missing. That is what you want, else process would not make any DBIO. That is not the cause the cause of the problem. If course if you do not perform any SQL INSERT the problem disappears. ;-)

**Comment by mpranjic** [comment]
> kwargs['dbio'] = kwargs.get('dbio', True)
> 
> only sets it true if missing. That is what you want, else process would not make any DBIO. That is not the cause the cause of the problem. If course if you do not perform any SQL INSERT the problem disappears. ;-)

Yeah, that line should be there, but ```dbio=False``` is never set for factory forms so it defaults to ```True```.
It somehow worked until now, perhaps because of passing ```DAL(None)``` in factory method.
Somewhere in the logic factory forms should prevent ```dbio``` implicitly.

[this line](https://github.com/web2py/web2py/blob/master/gluon/sqlhtml.py#L1688) gets called even for factory forms. But the change in DAL no longer "ignores" it.

Another question is, is the DAL behavior correct now? What is it trying to write into?
[```DAL(None)```](https://github.com/web2py/web2py/blob/master/gluon/sqlhtml.py#L1749) is not a valid connection / db backend :).

To summarize, I think both pydal and web2py need to be fixed.

**Comment by mdipierro** [comment]
I need to see your code. When we do

     form = SQLFORM.factory(Field('name'))

we never pass a db object to it, therefore if there is no db, there cannot be any dbio, not even accidentally.

**Comment by mpranjic** [comment]
we never pass the object, it creates one: [see here](https://github.com/web2py/web2py/blob/master/gluon/sqlhtml.py#L1749)

**Comment by mpranjic** [comment]
> I need to see your code

It's actually in the traceback, there's just this line:
```
form = SQLFORM.factory(Field('email', 'string'), Field('password', 'string')).process(keepvalues=True)
```
nothing unusual...

**Comment by mdipierro** [comment]
Damn. One minute ago I could not reproduce it, not I can. Fascinating!

**Comment by mdipierro** [comment]
1) One problem is that DAL behavior has changed and needs to be restored:
   db = DAL(None)
   db.define_table('a',Field('b'))
   db.a.insert(b='x')
The specs are that if uri = None, than is should use the DummyAdapter that implements insert, select, update, delete but simply ignores them. Since this is a change of behavior it should be reinstated.

2) Another problem is that the SQLFORM.factory behavior is stupid and it should skip the call to insert since there is no need to generate the query and then ignore it. 

@gi0baro if you could fix what has changed to cause 1), I will fix problem 2).

**Comment by gi0baro** [comment]
@mdipierro ok

**Comment by mdipierro** [comment]
Let me know when done

**Comment by mdipierro** [comment]
Should be fixed in branch master, please check it/
